sally/models/stock_lot.py

```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+
+from odoo import models, fields, api
+from odoo.exceptions import ValidationError
+
+
+class FlowerStockLot(models.Model):
+    _inherit = "stock.lot"
+
+    water_ids = fields.One2many("flower.water", "serial_id")
+    is_flower = fields.Boolean(
+        string='Is Flower Product?', related='product_id.is_flower')
+    needs_watering = fields.Boolean(
+        string='Needs Watering?', related='product_id.needs_watering')
+    user_ids = fields.Many2many('res.users', string='Gardner')
+
+    def action_water_flower(self):
+        for record in self.filtered(lambda rec: rec.is_flower):
+            today = fields.Date.today()
+            if record.water_ids:
+                last_watered_date = record.water_ids[0].date
+                frequency = record.product_id.flower_id.watering_frequency
+                if (today - last_watered_date).days < frequency:
+                    raise ValidationError(
+                        f"You shall not water this flower because it has been watered!\nYou may try again in {frequency-(today - last_watered_date).days} days.")
+            self.env["flower.water"].create({
+                "serial_id": record.id,
+                "date": today,
+                "gardner": self.env.user.id,
+                "rain_watering": False,
+            })
+        self.env['sally.flowers'].action_needs_watering()
```

Comment:
will work perfectly fine. However a better approach here would be storing the bad values in a list and raising an error after  committing all the creations.

The reason  behind is that errors in odoo will stop the execution of the program. meaning nothing will be committed to the database and any values changed or created records would be rolled back. So it will be a better approach to identify and store bad values in a list, creating the records if any, committing, then raising an error